Index: glibc-2.4/iconvdata/run-iconv-test.sh
===================================================================
--- glibc-2.4.orig/iconvdata/run-iconv-test.sh	2006-04-21 19:34:19.000000000 -0400
+++ glibc-2.4/iconvdata/run-iconv-test.sh	2006-04-21 19:34:58.000000000 -0400
@@ -128,7 +128,7 @@
     done
   fi
 
-  if test "$subset" == N; then
+  if test "$subset" = N; then
     echo $ac_n "      suntzu: ASCII -> $to -> ASCII $ac_c"
     $PROG -f ASCII -t $to testdata/suntzus |
     $PROG -f $to -t ASCII > $temp1 ||
Index: glibc-2.4.orig/linuxthreads/tst-tls2.sh
===================================================================
--- glibc-2.4/linuxthreads/tst-tls2.sh	2003-09-02 02:29:31.000000000 +0200
+++ glibc-2.4/linuxthreads/tst-tls2.sh	2006-08-20 16:28:09.430784935 +0200
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
 
 common_objpfx=$1; shift
 elf_objpfx=$1; shift
Index: glibc-2.4/nptl/tst-tls6.sh
===================================================================
--- glibc-2.4.orig/nptl/tst-tls6.sh	2003-09-03 00:02:59.000000000 +0200
+++ glibc-2.4/nptl/tst-tls6.sh	2006-08-20 16:28:31.045608966 +0200
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
 
 common_objpfx=$1; shift
 elf_objpfx=$1; shift
