From: Kees Cook <kees@ubuntu.com>
Subject: glibc $ORIGIN fix
Date: Thu, 21 Oct 2010 11:39:36 -0700
Organization: Ubuntu
Message-ID: <20101021183936.GI6311@outflux.net>

--- elf/rtld.c~	2010-10-21 11:37:39.857597987 -0700
+++ elf/rtld.c	2010-10-21 11:38:24.840276352 -0700
@@ -2488,7 +2488,7 @@
   while ((p = (strsep) (&str, ":")) != NULL)
     if (p[0] != '\0'
 	&& (__builtin_expect (! INTUSE(__libc_enable_secure), 1)
-	    || strchr (p, '/') == NULL))
+	))
       {
 	/* This is using the local malloc, not the system malloc.  The
 	   memory can never be freed.  */
