--- sysdeps/posix/getaddrinfo.c.orig	2007-10-24 18:27:00.000000000 +0200
+++ sysdeps/posix/getaddrinfo.c	2007-10-24 18:27:06.000000000 +0200
@@ -273,7 +273,7 @@
 static int
 gaih_inet (const char *name, const struct gaih_service *service,
 	   const struct addrinfo *req, struct addrinfo **pai,
-	   unsigned int *naddrs)
+	   unsigned int *naddrs, bool usable_ipv6)
 {
   const struct gaih_typeproto *tp = gaih_inet_typeproto;
   struct gaih_servtuple *st = (struct gaih_servtuple *) &nullserv;
@@ -716,7 +716,7 @@
 	      if (fct != NULL)
 		{
 		  if (req->ai_family == AF_INET6
-		      || req->ai_family == AF_UNSPEC)
+		      || (req->ai_family == AF_UNSPEC && usable_ipv6))
 		    {
 		      gethosts (AF_INET6, struct in6_addr);
 		      no_inet6_data = no_data;
@@ -1907,7 +1907,7 @@
   if (hints->ai_family == AF_UNSPEC || hints->ai_family == AF_INET
       || hints->ai_family == AF_INET6)
     {
-      last_i = gaih_inet (name, pservice, hints, end, &naddrs);
+      last_i = gaih_inet (name, pservice, hints, end, &naddrs, seen_ipv6);
       if (last_i != 0)
 	{
 	  freeaddrinfo (p);
--- sysdeps/unix/sysv/linux/check_pf.c.orig	2007-10-24 18:27:00.000000000 +0200
+++ sysdeps/unix/sysv/linux/check_pf.c	2007-10-24 18:27:06.000000000 +0200
@@ -151,7 +151,10 @@
 		  *seen_ipv4 = true;
 		  break;
 		case AF_INET6:
-		  *seen_ipv6 = true;
+		  if (ifam->ifa_scope < RT_SCOPE_LINK)
+		    {
+		      *seen_ipv6 = true;
+		    }
 
 		  if (ifam->ifa_flags & (IFA_F_DEPRECATED
 					 | IFA_F_TEMPORARY
