--- sysdeps/posix/getaddrinfo.c~	2008-07-30 19:14:22.000000000 +0000
+++ sysdeps/posix/getaddrinfo.c	2008-11-29 09:01:04.000000000 +0000
@@ -269,7 +269,7 @@
 static int
 gaih_inet (const char *name, const struct gaih_service *service,
 	   const struct addrinfo *req, struct addrinfo **pai,
-	   unsigned int *naddrs)
+	   unsigned int *naddrs, bool usable_ipv6)
 {
   const struct gaih_typeproto *tp = gaih_inet_typeproto;
   struct gaih_servtuple *st = (struct gaih_servtuple *) &nullserv;
@@ -762,7 +762,7 @@
 		  if (fct != NULL)
 		    {
 		      if (req->ai_family == AF_INET6
-			  || req->ai_family == AF_UNSPEC)
+			  || (req->ai_family == AF_UNSPEC && usable_ipv6))
 			{
 			  gethosts (AF_INET6, struct in6_addr);
 			  no_inet6_data = no_data;
@@ -2151,7 +2151,7 @@
   if (hints->ai_family == AF_UNSPEC || hints->ai_family == AF_INET
       || hints->ai_family == AF_INET6)
     {
-      last_i = gaih_inet (name, pservice, hints, end, &naddrs);
+      last_i = gaih_inet (name, pservice, hints, end, &naddrs, seen_ipv6);
       if (last_i != 0)
 	{
 	  freeaddrinfo (p);
--- sysdeps/unix/sysv/linux/check_pf.c~	2008-03-06 16:51:24.073172073 +0100
+++ sysdeps/unix/sysv/linux/check_pf.c	2008-03-06 17:08:00.978210159 +0100
@@ -178,7 +178,8 @@
 		    }
 		  else
 		    {
-		      if (!IN6_IS_ADDR_LOOPBACK (address))
+		      if (!IN6_IS_ADDR_LOOPBACK (address ?: local) &&
+			  ifam->ifa_scope < RT_SCOPE_LINK)
 			*seen_ipv6 = true;
 		    }
 		}
