Index: misc/regexp.h
===================================================================
--- misc/regexp.h.orig	2008-06-11 23:52:48.000000000 +0200
+++ misc/regexp.h	2008-06-11 23:54:41.000000000 +0200
@@ -129,8 +129,13 @@
   __expr_ptr = (regex_t *) expbuf;
   /* The remaining space in the buffer can be used for the compiled
      pattern.  */
+#ifdef __USE_GNU
   __expr_ptr->buffer = expbuf + sizeof (regex_t);
   __expr_ptr->allocated = endbuf -  (char *) __expr_ptr->buffer;
+#else
+  __expr_ptr->__buffer = expbuf + sizeof (regex_t);
+  __expr_ptr->__allocated = endbuf -  (char *) __expr_ptr->__buffer;
+#endif
 
   while ((__ch = (GETC ())) != eof)
     {
@@ -198,7 +203,11 @@
       }
 
   /* Everything is ok.  */
+#ifdef __USE_GNU
   RETURN ((char *) (__expr_ptr->buffer + __expr_ptr->used));
+#else
+  RETURN ((char *) (__expr_ptr->__buffer + __expr_ptr->used));
+#endif
 }
 #endif
 
