--- io/ftw.c.orig	2006-05-30 00:56:37.942174148 +0200
+++ io/ftw.c	2006-05-30 00:56:45.044815874 +0200
@@ -341,7 +341,7 @@
 
       if (dfdp != NULL && *dfdp != -1)
 	{
-	  int fd = openat64_not_cancel_3 (*dfdp, data->dirbuf + data->ftw.base,
+	  int fd = openat64 (*dfdp, data->dirbuf + data->ftw.base,
 					  O_RDONLY | O_DIRECTORY | O_NDELAY);
 	  dirp->stream = NULL;
 	  if (fd != -1 && (dirp->stream = __fdopendir (fd)) == NULL)
